### PR TITLE
Support experimental features

### LINKED ISSUE #6257: Allow running new features behind an experiments flag or variable
We want to work on new features that are right now ill-defined, like the [component testing](https://github.com/cypress-io/cypress/issues/5922). The best way to figure out the user experience and solve technical problems and unknowns is, in my opinion, to release this feature behind an "early beta experiment" flag, use it, collect feedback and iterate. Then we can release a fully developed feature without a flag.

## Proposal

List all experiments to enable in an environment variable, like

```
CYPRESS_EXPERIMENTS=componentTesting,alphaFeatureY=5,betaFeatureX=foo
```

The value of `CYPRESS_EXPERIMENTS` gets split by comma, then values are parsed, and the features should be set.

```js
{
  componentTesting: true,
  alphaFeatureY: 5,
  betaFeatureX: 'foo'
}
```

Note: we could grab this information from parsed Cypress config `Cypress.config('experiments')` during tests, and we can probably use a similar approach from CLI and `packages/*` code.

For each experiment, we should add a record in our docs, under redirect `on.cypress.io/experiments` describing the flag, the experiment, the linked issue, etc.

When using `cypress run`, if there are enabled experiments, put message into the run header

## Stretch goal

In the Test Runner Desktop GUI (the first window that pops up when you open Cypress) add another tab with experiments as checkboxes.

### PR DESCRIPTION
**Pull request by bahmutov**
- Closes #6257

### Additional details

- boolean flags only for now from `process.env.CYPRESS_EXPERIMENTS`. Unknown experiments are ignored
- trying it in action for component testing https://github.com/cypress-io/cypress/pull/5923
- on the Node side use `packages/server/lib/experiments.ts`

```js
// example
const experiments = require('./experiments').getExperiments()
if (experiments.componentTesting) {
 ...
}
```
- on the desktop GUI side, we grab the environment variable `EXPERIMENTS` from resolved config and parse it using the same logic from `packages/server`. Which means technically the experiment can be overwritten using plugins file or CLI for the browser and desktop GUI, while only getting value from the environment variable on the Node side

### How has the user experience changed?

There is a new tab and by default, since there is nothing whitelisted in the config, it is just notification text

Screen Shot 2020-02-07 at 4 56 28 PM

### PR Tasks

- [ ] have GUI and terminal changes been approved?

Then:
- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] On links updated and released for https://on.cypress.io/experiments

Plan is to merge this PR with `componentTesting` or empty list of experiments (?) and then merge https://github.com/cypress-io/cypress/pull/5923 with `componentTesting` experiment (and a few tests for new mounting)

### CODE REVIEW

**Comment by bahmutov** [comment]
Added Desktop GUI tab showing experiments

Screen Shot 2020-01-28 at 12 04 06 PM

**Comment by bahmutov** [comment]
## UI review

Desktop GUI experiments tab

"Learn more" link goes to (non-existent yet) https://on.cypress.io/experiments

### Experiment enabled

Screen Shot 2020-01-28 at 1 54 12 PM

### No experiments enabled

Screen Shot 2020-01-28 at 1 55 04 PM

- [ ] looks good, @brian-mann ?

**Comment by bahmutov** [comment]
## UI Review

Headless run with experiments enabled (without experiments same terminal output)

Screen Shot 2020-01-28 at 2 19 45 PM

- [ ] looks good, @brian-mann ?

File: packages/desktop-gui/src/settings/experiments.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        <pre className='line-nums'>
+          <span>CYPRESS_EXPERIMENTS=featureA,featureB cypress open</span>
+          <span># or</span>
+          <span>CYPRESS_EXPERIMENTS=featureA,featureB cypress run</span>
+        </pre>
+      </div>
+
+      <div>
+        <h3><input type="checkbox" readOnly defaultChecked={experiments.componentTesting} /> component testing</h3>
```
Comment (flotwig) [inline packages/desktop-gui/src/settings/experiments.jsx]: i don't know how crazy I am about using a disabled checkbox to distinguish between enabled and disabled - would prefer "Status: Enabled" or "Status: Disabled" instead, for example

Comment (bahmutov) [reply packages/desktop-gui/src/settings/experiments.jsx]: wait, something is weird, this is missing some of my commits!

Comment (bahmutov) [reply packages/desktop-gui/src/settings/experiments.jsx]: now it is good!

File: packages/desktop-gui/src/settings/experiments.jsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            <p>The environment variable currently has the value</p>
+              <pre className='line-nums'>
+                <span>CYPRESS_EXPERIMENTS={resolvedEnv}</span>
+              </pre>
+          </>
+        }
+      </div>
+
+      <div>
+        <h3>{experiments.componentTesting ? '✔️' : '🚫'} component testing</h3>
```
Comment (flotwig) [inline packages/desktop-gui/src/settings/experiments.jsx]: oh yeah, this is what i was talking about. i feel like the emoji are unclear, can we use "status: enabled/disabled" or something instead?

Comment (bahmutov) [reply packages/desktop-gui/src/settings/experiments.jsx]: look at the screenshots above, they are beautiful https://github.com/cypress-io/cypress/pull/6265#issuecomment-579402112

Comment (bahmutov) [reply packages/desktop-gui/src/settings/experiments.jsx]: but I am not against something like `status: enabled` or `status: disabled` labels instead.

Comment (flotwig) [reply packages/desktop-gui/src/settings/experiments.jsx]: yeah, I think that would be clearer

**Comment by jennifer-shehane** [comment]
I'm gonna work on the Desktop-GUI to freshen the design up for the experiements flags display.

**Comment by bahmutov** [comment]
## Feedback - changes requested

- instead of creating new wrapper and only accepting environment variable, put experiments into the existing config. That way the flags can be passed in `cypress.json`, CLI, environment variable or overwritten by the plugins. When you change the `cypress.json`, the test runner will reload, just like now, and would not require every user setting an environment variable.
- name experimental properties in the config starting with `experimental_`. This way the Experiments settings tab can still show them nicely

**Comment by brian-mann** [comment]
@bahmutov I would camel case the experiments...

`experimentalComponentTesting`
vs
`experimental_componentTesting`

**Comment by bahmutov** [comment]
Updated logic to just use `config`, still have to clean up code a little bit from previous code before pushing.

Experimental properties are just keys in the config

Screen Shot 2020-01-29 at 3 14 20 PM

The experiments tab in Desktop GUI picks those properties and shows them nicely if they are non-default

Screen Shot 2020-01-29 at 3 29 28 PM

**Comment by jennifer-shehane** [comment]
When we talked about this earlier, I assumed this would be a nested object, which seems cleaner to me than a bunch of very long config values starting with `experiemental`, but....I guess this complicates things when wanting to just turn it on with one env var.  

```
"experiments": {
  "componentTesting": true,
}
```

Looked at some other tools cause I wasn't totally satisfied.

- **Docker** seems to have you set `experimental: "enabled"`, and then just list the features you want out separately, which I don't think is great either. https://docs.docker.com/engine/reference/commandline/cli/#experimental-features
- **Node** does have everything prefixed with `experimental` like `--experimental-json-modules` https://nodejs.org/api/cli.html
- **Chrome** also has an `experimental` flag outright to turn them all on or off and then separate flags which are all prefixed `experimental`. https://developer.chrome.com/extensions/experimental
- **Kubernetes** uses `feature-gates` 😆 https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ Although they do have a nice feature where you can say, in the CLI print all of the experimental features available (probably overkill for us know though, they have 100's of features).
- Bazel has `experimental` prefix https://docs.bazel.build/versions/2.0.0/command-line-reference.html
- Parcel `experiemental` prefix.
- Everything else I saw doesn't explicitly specify 'experiemental' as part of actually setting the value and solely has it mentioned in documentation.


I think what looks weird is the camelCasing. Nobody camelCases the values but I get wanting to stick with our camelCase convention.

**Comment by jennifer-shehane** [comment]
Clicking into the Experiments panel completely borks if you do not have any experiments defined. I've updated the title to be `[WIP]` as this is not ready for review.

**Comment by jennifer-shehane** [comment]
My first iteration. I decided to pull inspiration from the Dashboard, tried to get a similar design for their Beta features - the main difference being they call theirs 'Cypress Labs' and ours is called 'Experiments'.

**Current Design**

Screen Shot 2020-01-30 at 2 49 47 PM


**Dashboard**

Screen Shot 2020-01-30 at 2 48 46 PM

**Comment by jennifer-shehane** [comment]
I started to write test in the Desktop Gui specs for the Experiments panel, but not nearly enough. I also updated the code to iterate through 'experiments' and not blow up when there are no experiment variables defined. 

This code is a bit naive. We won't be able to pull any key starting with `experimental`, we will have to explicitly whitelist all keys because people could just make up keys and it would do nothing. https://github.com/cypress-io/cypress/blob/experiments-6257/packages/server/lib/experiments.ts#L35:L35

We also need to build up more properties for each experiment for the display:

- title/name
- value
- default value
- description

**Comment by bahmutov** [comment]
@jennifer-shehane did you push any code? I don't see any commits. I love the UI with toggle - but these are read-only, right? since the user has to change them using `cypress.json` or CLI or env for now?

We still control the experiments, thus the user cannot just put _any_ `experimentalX` config value, we still whitelist keys, and only take the ones we know about.

As far as props for experiments: 'title/name', 'value', 'default value', 'description' - we have `value` coming from config object (resolved), title and description are the text in `experiments.jsx` and the only missing thing I think is `default value`. Unfortunately, since the experiment is part of config, our config has to have `resolved` objects with current `value` and `defaultValue` and `from` fields. We can tackle this in a separate PR - the config coffeescript code is already hairy enough I feel, I would 

- land this PR after more tweaks and tests
- rewrite config in TypeScript (at least the interface)
- add `default value` property to each resolved config value

**Comment by jennifer-shehane** [comment]
It is read only. They are not toggles, just display in the Desktop-GUI, although Brian suggested that they could be toggles (we just write the value directly to their `cypress.json`).

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
Sorry, maybe I was not clear in my explanation before. 

1. This line will `.map` through each experiment found:

https://github.com/cypress-io/cypress/blob/experiments-6257/packages/desktop-gui/src/settings/experiments.jsx#L28:L28

Right now the test for 'Component Testing' and 'Enables the use of ....' is statically written into the template. There needs to be a title and a description for each experimental feature so that when it loops through each future experiement this data can be shown. 

So an `experiment` should be an object constructed like the one below before it reaches the `.jsx` template:

```js
{
  title: "Component Testing"
  description: "Enables the use of component testing using framework-specific libraries to mount your component directly from the spec file."
  value: true
  default: false
}
```

2. This code is naive. It grabs any configuration value starting with `experimental`, which means if I pass the below, then `experimentalFooBar` will be shown in the Desktop Gui (but likely just crash the Desktop-GUI Settings panel in this case - and/or show itself under 'Component Testing' section due to #1 issue described above). 

We have to whitelist the experiment keys available. 

```js
{
  "chromeWebSecurity": false,
  "experimentalFooBar": true
}
```

3. There are inadequate tests in the Desktop Gui testing the display of the Component Testing toggle and other experimental keys being read in correctly. I added a single test just to get them started, but this is not testing that any of the content or the toggle is displaying correctly at all. https://github.com/cypress-io/cypress/pull/6265/files#diff-bb3b13ed2f43f3f1f9486a5593115e69R614

4. ~The `OFF` toggle shows as green.~ Fixed in https://github.com/cypress-io/cypress/commit/b5d1d87dd52bb23af93e5d7d3450504efcb8aeee

I'm not sure why `WIP` was removed from the title when neither of my previous comments from before about these things were addressed.

**Comment by bahmutov** [comment]
So, it is hard to work on this branch if the real experiment is in https://github.com/cypress-io/cypress/pull/5923 and I have updated that branch based on your UI (and removed `.map`, we need to be explicit)

Screen Shot 2020-02-03 at 8 25 13 AM

I also added Desktop GUI tests there, because it should set the experiment status. 

Ughh, this is what having two long-living pull requests means, neither is working very well. But to address your comments again:

- we are whitelisting config keys, thus unknown `experimental*` keys will be skipped anyway. We have to whitelist, because I suspect each `experimental*` key will also need some other `cypress.json` config keys. For example, component testing needs component tests folder setting.

So to move this forward, and land this pull request to actually do work for experimental component testing (that's the only thing that matters), I will keep your `.map` here, will move a few tests from https://github.com/cypress-io/cypress/pull/5923/commits/105b9c85ed9e6f97cb080e7be905cd9d160a7121 into this branch so at least we will know if desktop gui works. The real work will be in components testing PR

**Comment by bahmutov** [comment]
I have added more tests, added whitelist to ensure unknown experimental features do not show component testing experiment

Screen Shot 2020-02-03 at 9 01 13 AM

**Comment by jennifer-shehane** [comment]
@bahmutov Yeah, I'm not sure why you need the other repo at all. I never ran it with this PR, just mock everything in the tests.

**Review by jennifer-shehane — DISMISSED** [review]
@bahmutov Yeah, I'm not sure why you need the other repo at all. I never ran it with this PR, you can mock everything in the tests. 

You mentioned 'removing map', but maybe I'm not understanding what you are meaning. 

In this code in this PR, we are looping over the experiments in the JSX template and showing the exact same title and description for every single experiment that would ever be passed to it. 

https://github.com/cypress-io/cypress/blob/experiments-6257/packages/desktop-gui/src/settings/experiments.jsx#L33:L33

It is my understanding that this PR (NOT the component testing PR) is meant to support **any** experimental features that we deem necessary in the future. 

Right now the React template is static and will literally print 'Component Testing' and the description for component testing for any experimental feature added in the future. So I don't see this PR as meeting the basic function of supporting 'Experimental Features'.

**Comment by bahmutov** [comment]
There is nothing to be afraid - there is a whitelist of experiments right now, and only component testing is whitelisted (and this is in addition to whitelisting experimental keys in the config object)

```
// whilelist known experiments here. Note that "experimental*" keys are also
// whitelisted in the config code
const isKnownExperiment = (experiment, key) =>
  ['experimentalComponentTesting'].includes(key)
```

and this PR already has e2e tests with unknown and known experiments, https://github.com/cypress-io/cypress/pull/6265#issuecomment-581427919

In other words, nothing bad can happen, we should be good to go, it will be blank (just the experiments preamble text) once merged into develop.

**Comment by bahmutov** [comment]
@jennifer-shehane I have reworked the PR to be more to your liking - all experimental summaries and names are in a single place in server/lib/experiments.ts file and experiments.jsx is all dynamic and nothing is hardcoded.

**Comment by bahmutov** [comment]
- I have hidden the "Experiments" panel if there are no experiments
- the complex config objects - I think it makes sense to look at this in the components testing PR, since it will be a concrete question there

**Comment by jennifer-shehane** [comment]
Updated the styles of the ON/OFF switch, so new design up for feedback.

Screen Shot 2020-03-13 at 3 48 19 PM

Screen Shot 2020-03-13 at 3 49 02 PM

**Review by jennifer-shehane — COMMENTED** [review]
Overall this PR looks good to me, barring approval for the design changes on the 'Experiments' panel

File: packages/desktop-gui/src/settings/experiments.jsx [hunk 3]
```diff
@@ -22,30 +22,28 @@ const Experiments = observer(({ project }) => {
           If you'd like to try out what we're working on, you can enable these beta features for your project by setting configuration using the <code>experimental</code> prefix.
         </p>
       </div>
-      {!_.isEmpty(experiments) &&
```
Comment (bahmutov) [inline packages/desktop-gui/src/settings/experiments.jsx]: good catch

**Pull request merged**